File: bckup/0630/pso.py
import numpy as np
from config import *
import logging
from time import time

# TODO generate all random values at once to improve performance

def crossover(individual1, individual2, crossover_rate):
  """Resulting individual has part of the genes of both parents."""
  split_gene = int(np.random.uniform(0,len(individual1)-1))
  # before gene split_idx, genes are from individual1, after from
  # individual2
  return np.concatenate((individual1[:split_gene], individual2[split_gene:]))

def mutation(individual, mutation_rate, cluster):
  """Forces a mutation in a single gene."""
  # mutate one gene
  new_individual = individual[:]
  while 1:
    mutated_gene = int(np.random.uniform(0,len(individual)-1))
    if cluster[mutated_gene].sleep_prob != 0:
      if individual[mutated_gene]:
        new_individual[mutated_gene] = 0
      else:
        new_individual[mutated_gene] = 1
      break
  
  return new_individual

def fitness(individual, cluster):
  """Evaluates how fit is an individual."""
  sleep = []
  for idx, gene in enumerate(individual):
    if gene:
      sleep.append(cluster[idx])
  initial_energy    = cluster.initial_energy
  individual_energy = cluster.get_remaining_energy(sleep)
  total_coverage  = cluster.regions.initial_coverage
  total_overlapping = cluster.regions.initial_overlapping
  #time1 = time()
  individual_coverage, individual_overlapping = cluster.get_both(sleep)
  #time2 = time()
  #print(time2-time1)
  if initial_energy == 0:
    term1 = 0
  else:
    term1 = PSO_FITNESS_ALPHA*(1-individual_energy/initial_energy)

  if total_coverage == 0:
    term2 = 0
  else:
    term2 = PSO_FITNESS_BETA*(individual_coverage/total_coverage)

  if individual_overlapping == 0:
    term3 = 0
  else:
    term3 = PSO_FITNESS_GAMMA*(total_overlapping/individual_overlapping)

  return term1 + term2 + term3

def calculate_mutation_rate(iteration):
  return PSO_WMAX - (PSO_WMAX-PSO_WMIN)*iteration/PSO_MAX_ITERATIONS


def sleep_scheduling_pso(cluster):
  """Runs PSO to decide which nodes in the cluster will sleep. The cur-
  rent cluster head should not be put to sleep, otherwise all informa-
  tion for that node is lost.
  """
  logging.info('running sleep scheduling')
  cluster_size = len(cluster)

  # calculate sleep probability for each node
  cluster.calculate_sleep_prob()
  sleep_prob = [node.sleep_prob for node in cluster]

  # paint grid and generate regions
  cluster.paint_grid()

  # initialize particles, best local and best global
  particles  = []
  best_locals = []
  best_locals_fit = []
  best_global = np.zeros((cluster_size), dtype=np.uint8)
  best_global_fit = 0.0
  # a single particle is a array with 0 (sleep) or 1 (awake) for every
  # node in the cluster
  for idx_particle in range(0, PSO_NB_PARTICLES):
    # each particle contains information about the state of all nodes
    # in this cluster, indicating if they are active (1) /inactive (0)
    random = np.random.uniform(0, 1, (cluster_size))
    particle = random > sleep_prob
    best_local = particle[:]

    if idx_particle == 0:
      best_global = best_local[:]
      best_global_fit = fitness(best_global, cluster)

    best_locals_fit.append(fitness(best_local, cluster))
    if best_locals_fit[idx_particle] > best_global_fit:
      best_global = best_local[:]
      best_global_fit = best_locals_fit[idx_particle]

    particles.append(particle)      
    best_locals.append(best_local)

  shape_randoms = (PSO_MAX_ITERATIONS, PSO_NB_PARTICLES, 3)
  randoms = np.random.uniform(0, 1, shape_randoms)
  for it in range(0, PSO_MAX_ITERATIONS):
    logging.debug("pso iteration %d" % (it))
    for idx_particle in range(0, PSO_NB_PARTICLES):
      particle = particles[idx_particle]
      best_local = best_locals[idx_particle]

      mutation_rate   = calculate_mutation_rate(it)
      crossover_rate1 = 1 - it/PSO_MAX_ITERATIONS
      crossover_rate2 = 1 - crossover_rate1

      new_particle = particle[:]
      if randoms[it][idx_particle][0] > mutation_rate:
        new_particle = mutation(new_particle, mutation_rate, cluster)
      if randoms[it][idx_particle][1] > crossover_rate1:
        new_particle = crossover(new_particle, best_local, crossover_rate1)
      if randoms[it][idx_particle][2] > crossover_rate2:
        new_particle = crossover(new_particle, best_global,crossover_rate2)

      if should_update_particle(new_particle, best_global):
        particle = new_particle[:]

      # update best locals and best global
      particle_fitness = fitness(particle, cluster)
      if particle_fitness > best_locals_fit[idx_particle]:
        #print('found best_local')
        best_local = particle[:]
        best_locals_fit[idx_particle] = particle_fitness
      if particle_fitness > best_global_fit:
        #print('found best_global')
        best_global = particle[:]
        best_global_fit = particle_fitness

  # actually put nodes to sleep
  for idx, node in enumerate(cluster):
    node.is_sleeping = best_global[idx]

def should_update_particle(particle1, particle2):
  differences = particle1 == particle2
  nb_differences = np.count_nonzero(differences)
  if nb_differences > len(particle1)/2.0:
    return 1
  else:
    return 0

File: bckup/0704/pso.cpp
#include "pso.h"
#include <random>
#include <stdio.h>

PSO::PSO(dict_t exclusive, regions_t overlapping,
         vector<int> ids, u_int nb_particles)
    :_ids(ids),
     _nb_particles(nb_particles) {
  _regions = new Regions(exclusive, overlapping);

  _nb_nodes = ids.size();
  // initialize particles structure
  _particles   = vector<individual_t>(nb_particles, individual_t(_nb_nodes, 0));
  _best_locals = vector<individual_t>(nb_particles, individual_t(_nb_nodes, 0));
  _best_global = individual_t(_nb_nodes, 0);
  _best_local_fitness = vector<float>(nb_particles, 0.0);
}

PSO::~PSO() {
  delete _regions;
}

individual_t
PSO::run(vector<float> energies, vector<float> sleep_probs) {
  default_random_engine generator;
  uniform_real_distribution<float> distribution(0.0, 1.0);

  float total_energy = 0.0;
  for (auto const &energy: energies) {
    total_energy += energy;
  }

  // initialize data
  _best_global_fitness = 0.0;
  //printf("number of particles %d\n", _nb_particles);
  for (u_int particle_idx = 0; particle_idx < _nb_particles; particle_idx++) {
    individual_t particle = _particles[particle_idx];
    for (u_int node_idx = 0; node_idx < _nb_nodes; node_idx++) {
      float random = distribution(generator);
      particle[node_idx] = (random < sleep_probs[node_idx]) ? 1 : 0;
    } 
    _best_locals[particle_idx] = particle;

    _best_local_fitness[particle_idx] = fitness(particle,
                                                energies, total_energy);

    if ((particle_idx == 0) ||
        (_best_local_fitness[particle_idx] > _best_global_fitness)) {
      _best_global = particle;
      _best_global_fitness = _best_local_fitness[particle_idx];
    }
  }
  // search for solution
  float mutation_rate, crossover_rate1, crossover_rate2;
  for (u_int it = 0; it < 50; it++) {
    for (u_int particle_idx = 0; particle_idx < _nb_particles; particle_idx++) {
      individual_t &particle = _particles[particle_idx];
      mutation_rate   = 0.9 - (0.9-0.4)*it/50.0;  
      crossover_rate1 = 1.0 - it/50.0;
      crossover_rate2 = 1.0 - crossover_rate1;

      individual_t new_particle = particle;
      if (distribution(generator) < mutation_rate) {
        //printf("mutating particle\n");
        new_particle = mutation(particle, sleep_probs);
      }
      if (distribution(generator) < crossover_rate1) {
        //printf("crossing1 particle\n");
        new_particle = crossover(particle, _best_locals[particle_idx]);
      }
      if (distribution(generator) < crossover_rate2) {
        //printf("crossing2 particle\n");
        new_particle = crossover(particle, _best_global);
      }
      if (distribution(generator) < should_update(new_particle, _best_global)) {
        //printf("updated particle\n");
        particle = new_particle;

        float particle_fitness = fitness(particle,
                                         energies, total_energy);
        if (particle_fitness > _best_local_fitness[particle_idx]) {
          _best_locals[particle_idx] = particle;
          _best_local_fitness[particle_idx] = particle_fitness;
        }
        if (particle_fitness > _best_global_fitness) {
          _best_global = particle;
          _best_global_fitness = particle_fitness;
        }
      }
    }
  }

  //printf("%f\n", _best_global_fitness);
  return _best_global;
}

individual_t
PSO::mutation(individual_t individual, vector<float> sleep_probs) {
  individual_t new_individual = individual;
  float threshold = 0.005;
  // find all nodes that are susceptible to sleep (not dead neither ch)
  vector<u_int> can_sleep;
  for(u_int idx = 0; idx < sleep_probs.size(); idx++) {
    float prob = sleep_probs[idx];
    if (prob > threshold) {
      can_sleep.push_back(idx);
    }
  }
  if (can_sleep.size() > 0) {
    default_random_engine generator;
    uniform_int_distribution<int> distribution(0, can_sleep.size());
     while (1) {
      u_int mutated_gene = distribution(generator);
      if (individual[mutated_gene] == 1) {
        new_individual[mutated_gene] = 0;
      } else {
        new_individual[mutated_gene] = 1;
      }
      break;
    }
  }
  return new_individual;
}

individual_t
PSO::crossover(individual_t individual1, individual_t individual2) {
  default_random_engine generator;
  uniform_int_distribution<int> distribution(0, individual1.size());

  individual_t new_individual = individual1;
  int split_gene = distribution(generator);
  for (u_int gene = split_gene; gene < individual1.size(); gene++) {
    new_individual[gene] = individual2[gene];
  }
  return new_individual;
}

float
PSO::should_update(individual_t particle1, individual_t particle2) {
  float nb_differences = 0.0;
  for (u_int gene = 0; gene < particle1.size(); gene++) {
    //printf("%d", particle1[gene]);
    //printf("%d", particle2[gene]);
    if (particle1[gene] != particle2[gene]) {
      nb_differences += 1.0;
    }
  }
  return nb_differences/float(particle1.size());
}

float
PSO::fitness(individual_t individual, vector<float> energies,
             float total_energy) {
  vector<int> sleep_nodes;
  float partial_energy = 0.0;
  for (u_int gene_idx = 0; gene_idx < _nb_nodes; gene_idx++) {
    if (individual[gene_idx] == 1) {
      sleep_nodes.push_back(_ids[gene_idx]);
    } else {
      partial_energy += energies[gene_idx];
    }
  }

  vector<float> coverage_info = _regions->get_all(sleep_nodes);

  float term1;
  if (total_energy == 0) {
    term1 = 0.0;
  } else {
    term1 = 0.3*(1-partial_energy/total_energy);
  }
  float term2;
  if (coverage_info[0] == 0) {
    term2 = 0.0;
  } else {
    term2 = 0.3*(coverage_info[1]/coverage_info[0]);
  }
  float term3;
  if (coverage_info[3] == 0) {
    term3 = 0.0;
  } else {
    term3 = 0.4*(coverage_info[2]/coverage_info[3]);
  }

  //printf("fitness %f\n", term1 + term2 + term3);
  return term1 + term2 + term3;
}


